### PR TITLE
ENH: Make tensorflow,torch,zarr packages optional

### LINKED ISSUE #134: Make `torch` and `tensorflow` be optional installation dependencies
We often use `histomics_stream` for its ability to mask whole slide images and/or to group (chunk) tiles into more-efficient reads.  Neither of these functionalities should need `tensorflow` or `torch`, so we should provide these functionalities without requiring installation of these two packages.

### PR DESCRIPTION
**Pull request by Leengit**
Closes #134 .

### CODE REVIEW

**Comment by Leengit** [comment]
The tests are failing because `pooch` can no longer find the test image and its mask.  @cooperlab have they been deleted or moved?:
```python
# download image
wsi_path = pooch.retrieve(
    fname="TCGA-AN-A0G0-01Z-00-DX1.svs",
    url=(
        "https://drive.google.com/uc"
        "?export=download"
        "&id=19agE_0cWY582szhOVxp9h3kozRfB4CvV"
        "&confirm=t"
        "&uuid=6f2d51e7-9366-4e98-abc7-4f77427dd02c"
        "&at=ALgDtswlqJJw1KU7P3Z1tZNcE01I:1679111148632"
    ),
    known_hash="d046f952759ff6987374786768fc588740eef1e54e4e295a684f3bd356c8528f",
    path=str(pooch.os_cache("pooch")) + os.sep + "wsi",
)

# download binary mask image
mask_path = pooch.retrieve(
    fname="TCGA-AN-A0G0-01Z-00-DX1.mask.png",
    url=(
        "https://drive.google.com/uc"
        "?export=download"
        "&id=17GOOHbL8Bo3933rdIui82akr7stbRfta"
    ),
    known_hash="bb657ead9fd3b8284db6ecc1ca8a1efa57a0e9fd73d2ea63ce6053fbd3d65171",
    path=str(pooch.os_cache("pooch")) + os.sep + "wsi",
)
print(f"Have {mask_path}")
```

**Comment by cooperlab** [comment]
@andsild can fix this for us

**Comment by andsild** [comment]
There are some differences in the URL, not sure why they changed, but here's what we use (changing the URL should be sufficient):
```python
wsi_fname = "TCGA-AN-A0G0-01Z-00-DX1.svs"
    wsi_path = pooch.retrieve(
        fname=wsi_fname,
        url="https://drive.usercontent.google.com/download?id=19agE_0cWY582szhOVxp9h3kozRfB4CvV&export=download&confirm=t",
        known_hash="d046f952759ff6987374786768fc588740eef1e54e4e295a684f3bd356c8528f",
        path=wsi_files,
    )
    print(f"Downloaded {wsi_fname} to {wsi_files}")

    mask_fname = "TCGA-AN-A0G0-01Z-00-DX1.mask.png"
    mask_path = pooch.retrieve(
        fname=mask_fname,
        url="https://drive.usercontent.google.com/download?id=17GOOHbL8Bo3933rdIui82akr7stbRfta&export=download&confirm=t",
        known_hash="bb657ead9fd3b8284db6ecc1ca8a1efa57a0e9fd73d2ea63ce6053fbd3d65171",
        path=wsi_files,
    )
    print(f"Downloaded {mask_fname} to {wsi_files}")
```

**Comment by andsild** [comment]
(if you want me to I could fix it, but I dont have write privileges in this repository)

**Comment by Leengit** [comment]
Thank you for the updated URLs.  I have made the changes myself, and have given you permissions for the repository.

**Comment by Leengit** [comment]
With the current set of commits, on GitHub only, with Python 3.8, somewhere deep inside a dependency of `large_image`, and regardless of various permutations and combinations of `pip install typing-extensions` and placement of `import large_image` statements that I have tried, I am getting
```text
ImportError: cannot import name 'Buffer' from 'typing_extensions' (/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/typing_extensions.py)
```
@manthey et al. have you seen this before?  It works fine locally for me.  It works fine with Python 3.6 and 3.7.

**Comment by Leengit** [comment]
Because it builds on my local system with Python 3.8, and it builds on GitHub with Python 3.6, 3.7, and 3.9, I have gone forward with the hack response of removing the Python 3.8 test build from GitHub.  If approved and merged as is, we should submit an issue to find a fix for the Python 3.8 build on GitHub.

**Comment by Leengit** [comment]
This pull request is ready for review.

**Comment by manthey** [comment]
> With the current set of commits, on GitHub only, with Python 3.8 only, somewhere deep inside a dependency of `large_image`, and regardless of various permutations and combinations of `pip install typing-extensions` and placement of `import large_image` statements that I have tried, I am getting
> 
> ```
> ImportError: cannot import name 'Buffer' from 'typing_extensions' (/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/typing_extensions.py)
> ```
> 
> @manthey et al. have you seen this before? Python 3.8 works fine locally for me. It works fine with Python 3.6, 3.7, and 3.9.

I expect that this means some other package is specifying typing_extensions < 4.6 in the 3.8 environment, since `Buffer` was only added then.

**Comment by manthey** [comment]
> > With the current set of commits, on GitHub only, with Python 3.8 only, somewhere deep inside a dependency of `large_image`, and regardless of various permutations and combinations of `pip install typing-extensions` and placement of `import large_image` statements that I have tried, I am getting
> > ```
> > ImportError: cannot import name 'Buffer' from 'typing_extensions' (/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/typing_extensions.py)
> > ```
> >   
> > @manthey et al. have you seen this before? Python 3.8 works fine locally for me. It works fine with Python 3.6, 3.7, and 3.9.
> 
> I expect that this means some other package is specifying typing_extensions < 4.6 in the 3.8 environment, since `Buffer` was only added then.

And, looking further, the version of TensorFlow that installs on 3.8 (version 2.13.1) requires typing_extensions< 4.6, so TensorFlow on 3.8 is incompatible with large_image on 3.8 because of that.  If we _want_ these to be compatible, we'd have to relax the type rules in large_image.

File: .github/workflows/build-test-package.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ jobs:
     strategy:
       max-parallel: 2
       matrix:
-        python-version: ["3.6", "3.7", "3.8", "3.9"]
+        python-version: ["3.6", "3.7", "3.9"]
```
Comment (manthey) [inline .github/workflows/build-test-package.yml:11]: Is there are reason we test on 3.6 and 3.7 at this point?  Should we instead test on something more modern?

File: .github/workflows/build-test-package.yml [hunk 2]
```diff
@@ -37,11 +37,12 @@ jobs:
           sudo apt install openslide-tools python3-openslide
           python -m pip install --upgrade pip setuptools wheel
           pip install 'flake8${{ matrix.flake8-python-git-tag }}' 'pooch${{ matrix.pooch-python-git-tag }}' 'pytest${{ matrix.pytest-python-git-tag }}'
-          pip install 'large-image[bioformats,ometiff,openjpeg,openslide,tiff]' 'scikit_image' --find-links https://girder.github.io/large_image_wheels
+          pip install --upgrade 'large-image[bioformats,ometiff,openjpeg,openslide,tiff]' 'scikit_image' --find-links https://girder.github.io/large_image_wheels
+          pip install --upgrade typing-extensions
```
Comment (manthey) [inline .github/workflows/build-test-package.yml:41]: Probably if you moved this after installing tensorflow it would work on Python 3.8.

**Review by manthey — APPROVED** [review]
This looks fine, though we could further adjust tests for python versions.

File: .github/workflows/build-test-package.yml (hunk 1 above)
Comment (Leengit) [reply .github/workflows/build-test-package.yml]: I just pushed a commit to use versions 3.8-3.12, which is 3.8 plus the [current tensorflow specifications](https://www.tensorflow.org/install/pip).  I also moved the `pip install --upgrade typing-extensions` command in `build-test-package.yml` to be after the tensorflow install that occurs with the histomics_stream install command, `pip install .[tensorflow,torch,zarr]`.  If Python 3.8 fails, I will remove it again.

File: .github/workflows/build-test-package.yml (hunk 2 above)
Comment (Leengit) [reply .github/workflows/build-test-package.yml]: Yes, that works.

File: .github/workflows/build-test-package.yml (hunk 1 above)
Comment (Leengit) [reply .github/workflows/build-test-package.yml]: Torch doesn't like Python 3.12, so I've changed the range to by Python 3.8-3.11.

**Pull request merged**